[클린코드 리액트 3기 이준근] 페이먼츠 미션 Step1

Jxxunnn opened the pull request:
## 프로젝트 실행 방법
프로젝트 시작하기
```bash
npm install && npm run dev
```
스토리북 시작하기
```bash
npm run storybook
```

## 기술 스택

- React
- TypeScript
- Tailwind CSS
- Storybook
- Vite

## 작업 내용

필수 요구사항
- [x] 원시적인 형태의 Primitive UI 형태의 컴포넌트 작성
- [x] Funnel 기반의 애플리케이션 설계
- [x] Storybook 상호 작용 테스트
- [x] Controlled & Uncontrolled Components를 명확하게 구분하거나 선택하여 구현

카드 추가
- [x] <(뒤로가기) 버튼 클릭 시, 카드 목록 페이지로 이동한다.
- [x] 카드 번호를 입력 받을 수 있다.
- [x] 만료일을 입력 받을 수 있다.
- [x] 보안코드를 입력 받을 수 있다.
- [x] 카드 비밀번호의 앞 2자리를 입력 받을 수 있다.
- [x] 카드 소유자 이름을 입력 받을 수 있다.
- [x] 카드 추가 완료시 카드 등록 완료 페이지로 이동한다.

## 테스트

- **스토리북 구성:** UI 컴포넌트에 대한 스토리를 작성했습니다.
- **e2e 테스트 세팅:** Funnel 기반 multi-step form 개발을 대비하여 e2e 테스트 환경을 설정했습니다.

## 기타 사항

### ❗ 아토믹 디자인 패턴 적용하였습니다. 아토믹 디자인 단위를 나누는 기준의 주관성은 다음과 같습니다. 
    - atom은 더 이상 ui적으로 분해할 수 없는 기본 컴포넌트
    - molcule은 여러 개의 atom, molcule을 조합하여 구성, 
    - organism은 도메인 컴포넌트로 여러 개의 atom, molcule, organism을 조합하여 구성, 
    - template은 페이지의 모든 컨텐츠 

### ❗ 필수 요구사항에 집중했습니다.

해당 미션의 source of truth인 피그마를 바탕으로 이해해야하는 도메인 컨텍스트가 추상적이고 이에 대해 논의할 사람이 없는 상황에서 primitive ui를 구축하고 이를 기반으로 기능 요구사항을 모두 충족시키는 도메인 컴포넌트를 쌓아가는 것이 너무 어려웠어요 ㅠ 

공통 UI같은 경우에도 서비스의 전역적인 도메인을 담은 디자인 시스템을 기반으로 작업하는 것이 아닌지라 확장성있고 유연한 컴포넌트로 바텀업 빌드하는 작업이 창작의 영역으로 느껴졌습니다. 🥲 

## 질문

### ❓ 아토믹 디자인 패턴에서 template과 page를 어떻게 구분짓는 것이 적절할까요?

보통 template과 page를 데이터 바인딩이 되었는지로 구분짓는 것 같은데, 리액트에서 해당 패턴을 적용한다는 것이 container-presenter 패턴을 의미하는 것은 아닐 것 같아요.

아토믹 디자인 패턴을 실무에 적용해보셨다면 어떠한 기준으로 계층을 나누셨고 어떤 고민과 문제가 있으셨는지 자세하게 듣고 싶습니다!

### ❓ CDD란 무엇인가요?

아토믹 디자인 패턴을 활용한 CDD같은 경우에는 회사의 도메인 컨텍스트를 고려하면서 아토믹 방법론에 입각하여 디자인을 기본 단위 인터페이스로 분해하고, 리액트스럽게 컴포넌트를 설계한 후 결합하는 과정이 있다고 생각하는데요. 

리뷰어님이 생각하시는 CDD란 무엇인지와 적용하셨을 때 느끼셨던 장단점이 궁금합니다.

### ❓ 리뷰는 어떻게 하는 것이 좋을까요?

리뷰어님이 평소 코드리뷰하시는 노하우가 궁금합니다! 어떻게 하면 리뷰를 잘 받을 수 있을지, 어떻게 하면 리뷰를 잘 할 수 있을지 모르겠어요.

### ❓ 스토리북을 평소 어떻게 활용하시나요?

스토리북은 사용법이 정말 무궁무진한 것 같아요. 현업에서 작업하실 때 스토리북은 워크플로우에서 어느 부분에 해당하는지, 어떤 기능까지 활용하시는지, 어느 부분이 장단점이라 느껴지셨는지, 협업 시에는 어떻게 활용하시며 어느 직군까지 스토리북으로 소통하시는지 등 스토리북을 적용하신 경험과 노하우가 궁금합니다!

### ❓ Funnel 기반으로 작성될 어플리케이션에서는 무엇을 고려해야할까요?

처음 이 요구 사항을 보고는 단순히 multi-stem from 에 대한 처리 방법으로 이해했지만, 주어가 '특정 폼 페이지'가 아니라 '어플리케이션'이라면 제가 생각한 것과 많이 다를 것 같습니다. 리뷰어님의 고견 듣고 싶습니다.

### ❓ 제어 컴포넌트, 비제어 컴포넌트

리뷰어님이 생각하시는 비제어 컴포넌트의 장점은 무엇인가요?

일단 저는 리액트에서는 제어 컴포넌트로 선개발하는 것이 유리하다고 생각합니다. 현대의 웹은 다양한 인터랙션과 그에 따른 피드백을 요구하고 UX요구사항도 빈번하여 비제어 컴포넌트는 이후 교체될 확률이 높다고 생각하거든요. 

해당 미션 같은 경우에도 차주 요구사항에 인풋 에러 메세지를 보여주는 기능이 추가되기에 비제어 컴포넌트로 개발되었다면 제어 컴포넌트로 교체되어야 할 것이라 예상하고 있습니다. 

제어 컴포넌트의 단점으로 대표적으로 언급되는 렌더링 성능 이슈는 리액트의 API로 충분히 방어할 수 있다고 생각해서 리액트에서 컴포넌트를 구현할 때 제어 vs 비제어를 고려해야하는 상황은 어느 때인지 잘 모르겠습니다. 리뷰어님의 고견 듣고 싶습니다!

Jxxunnn committed:
project kick-off: vite + react + ts

Jxxunnn committed:
chore: install and configure tailwind css

Jxxunnn committed:
feat: remove unused style files

Jxxunnn committed:
chore: install libraries

- react-router v6
- tailwind merge

Jxxunnn committed:
feat: config absolute path

Jxxunnn committed:
fix: style global css

Jxxunnn committed:
feat: implement react router

Jxxunnn committed:
feat: implement atom components

- Box
- Button
- Container
- HFlex
- Icon
- MobilePanel
- Portal
- Text
- VFlex

Jxxunnn committed:
feat: implement molecule components

- Header
- IconButton
- Input
- Modal

Jxxunnn committed:
feat: config typography variables

Jxxunnn committed:
feat: implement organism component

- CreditCardInput

Jxxunnn committed:
feat: implement template component

- CreateCardForm

Jxxunnn committed:
chore: add gitignore ongoing works

Jxxunnn committed:
feat: wrap the app with layout component

Jxxunnn committed:
chore: install and configure storybook

Jxxunnn committed:
feat: implement atom component stories

Jxxunnn committed:
feat: implement molecule component stories

Jxxunnn committed:
feat: implement organism component stories

Jxxunnn committed:
feat: implement template component story

Jxxunnn committed:
chore: temporarily removed routing logic

해당 페이지는 현재 스코프에 개발되지 않아, 임시로 alert로 대체하였습니다.

Jxxunnn committed:
feat: write storybook readme

igy95 commented on README.md:
`README`가 앱에 대한 설명에서 boiler plate template에 대한 설명으로 바뀐 듯 한데, 이건 유지하는 게 좋지 않을까요~?

Jxxunnn replied:
네 맞습니다. vite를 설치하면서 오버라이드된 부분을 신경쓰지 못했네요. 기존 리드미에 수정하였습니다. 🙏

Jxxunnn replied:
https://github.com/next-step/react-payments/pull/107/commits/77fee409888f199ec8541b7cafeb9b6172194ca6

igy95 commented on src/stories/READMe.mdx:
파일명 `README`로 case 맞춰줘야 할 것 같아요!

Jxxunnn replied:
좋은 피드백 감사합니다. 해당 파일은 스토리북을 활용하기 전에 살펴보면 좋은 정보를 담는 파일로 목적이 있어, readme라고 이름을 지었는데요, 말씀 주신 대로 case도 안맞고, 관례적으로 README 파일은 프로젝트에 한 개만 있다는 점을 고려하여 Description으로 파일 이름을 변경하였습니다.

Jxxunnn replied:
https://github.com/next-step/react-payments/pull/107/commits/defbb645e7a5166b377b608662a1601f8e35ae31

igy95 commented on .gitignore:
```diff
@@ -0,0 +1,29 @@
+*.local
+
+# Editor directories and files
+.vscode/*
```
이거는 왜 `.gitignore`에 추가하셨는지 궁금합니다!

Jxxunnn replied:
> 네 맞습니다. vite를 설치하면서 오버라이드된 부분을 신경쓰지 못했네요. 기존 리드미에 수정하였습니다. 🙏

동일한 부분인데요. 이러한 부분이 리뷰어님을 혼란스럽게 만든다는 점을 간과하고 꼼꼼하게 신경쓰지 못했네요. 죄송합니다. 
프로젝트 IDE 세팅에 관한 부분이고 현 프로젝트에서 필요하지 않으므로 삭제 처리 하겠습니다. 🥲

Jxxunnn replied:
https://github.com/next-step/react-payments/pull/107/commits/25339c87871f38721d4f9b368c459fdb41172bda

igy95 commented on src/App.tsx:
```diff
@@ -0,0 +1,16 @@
+    <Router>
+      <Routes>
+        <Route path="/" element={<CreateCardForm />} />
+        <Route path="/credit-card" element={<CreateCardForm />} />
+        <Route path="/credit-card/create" element={<CreateCardForm />} />
```
`element`가 동일해보이는데 path를 따로 두신 이유가 있을까요?

Jxxunnn replied:
라우팅 테스트 목적이었고 원격 저장소에 push될 이유는 없습니다 . 삭제 처리 하겠습니다. 🙏

Jxxunnn replied:
https://github.com/next-step/react-payments/pull/107/commits/a880e85f1f54c5bf7104ce25bb4bedaa8e96521a

igy95 commented on src/components/atoms/Box/index.tsx:
```diff
@@ -0,0 +1,9 @@
+import { ComponentProps } from 'react';
+
+export default function HFlex({ children, className, ...restProps }: ComponentProps<'div'>) {
```
- 파일명과 컴포넌트 이름을 맞춰줘야 할 것 같아요!
- 이걸 쓰는 것과 그냥 div를 쓰는 것이 별 차이가 없어 보이는데, 이 컴포넌트의 역할은 무엇인가요?

Jxxunnn replied:
> 파일명과 컴포넌트 이름을 맞춰줘야 할 것 같아요!

정확한 지적 감사합니다. 수정하겠습니다. 🙏
https://github.com/next-step/react-payments/pull/107/commits/008a1e4ff4cf5b3c698f3ec2c93f7d58c1314a01

> 이걸 쓰는 것과 그냥 div를 쓰는 것이 별 차이가 없어 보이는데, 이 컴포넌트의 역할은 무엇인가요?

1. 말씀해주신 것처럼 div와 같이 특별한 의미가 없는 컨테이너 역할로 활용하였습니다.

2. 컴포넌트화 한 이유는 필수 요구사항에 따라 primitive ui형태의 컴포넌트를 적극적으로 구현하면서 UI 코드베이스 내의 추상화 레벨을 맞추기 위함이었습니다.

igy95 replied:
> 컴포넌트화 한 이유는 필수 요구사항에 따라 primitive ui형태의 컴포넌트를 적극적으로 구현하면서 UI 코드베이스 내의 추상화 레벨을 맞추기 위함이었습니다.

자체 구현을 해야하는 시점은, 내가 지금 맞닥뜨린 문제를 해결하는 데 있어 기존에 있는 방법론 중 마땅한 방법이 없을 때라고 생각합니다. 즉 이미 만들어진 것과 새로 만든 것을 비교했을 때 뚜렷한 이점이 없다면, 오히려 사용할 때 혼동을 줄 수 있는 부분이지 않을까 싶어요.

Jxxunnn replied:
> 자체 구현을 해야하는 시점은, 내가 지금 맞닥뜨린 문제를 해결하는 데 있어 기존에 있는 방법론 중 마땅한 방법이 없을 때

마음에 새기며 개발에 임하겠습니다. 감사합니다. 🙏

igy95 commented on src/components/atoms/Container/index.tsx:
`Box`, `Container`라는 이름이 꽤 포괄적으로 느껴져 어떤 상황에서 사용해야 할지 감이 잘 안옵니다!

Jxxunnn replied:
사용법은 스토리북의 mdx파일에 작성하였고, 이름은 MUI에서 영감을 얻었습니다.
https://mui.com/material-ui/react-box/
https://mui.com/material-ui/react-container/

```
# Box

Box 컴포넌트는 다목적 컨테이너 역할을 하는 기본적인 빌딩 블록입니다.
```
```
# Container

Container 컴포넌트는 페이지의 레이아웃을 구성하는데 사용됩니다.
```

primitive ui 컴포넌트와 추상화 레벨을 맞추기 위한 일련의 작업이었고, Box 컴포넌트는 다른  div와 span같은 meaningless element를 대체, Container 컴포넌트는 main, section ,article같은 sectioning element를 대체하고자 만들었습니다.

각자의 요소에서 자주 쓰이는 css property를 prop으로 받아 직관성을 높이는 방향으로 확장시킬 생각이었는데요. 이번 스텝에서는 불필요한 확장이라 생각이 들어 진행하지 않았습니다. 

컴포넌트 이름을 시맨틱하게 사용하여 가독성을 향상시키고자 한 시도라고 봐주셔도 무방할 것 같습니다. 🙏

igy95 replied:
네 이해했습니다!

igy95 commented on src/components/atoms/Portal/index.tsx:
```diff
@@ -0,0 +1,24 @@
+  );
+}
+
+// Fnull
```
필요없는 주석은 지워주세요~

Jxxunnn replied:
넵🙏

https://github.com/next-step/react-payments/pull/107/commits/d8c1538b836c56717f0da5df4d91b02e3c9d3dde

igy95 commented on src/styles/guide.ts:
```diff
@@ -0,0 +1,10 @@
+export const TYPOGRAPHY_SIZE_MAP = {
```
`TYPOGRAPHY_SIZE_MAP` 내의 key와 value도 각각의 타입을 가지고 있어야 이 코드를 읽는 입장에서도, 사용하는 입장에서도 타입스크립트의 이점을 좀 더 가져갈 수 있을 것 같습니다. (하단 예시는 pseudo code이니 예시로만 봐주세요!)

```ts
export const TYPOGRAPHY_SIZE_MAP: Record<TypographySemanticName, TailwindValue> = {}
```

Jxxunnn replied:
https://github.com/next-step/react-payments/pull/107/commits/74393418c5e03e86ec813c6672e0c0ef9ee1ac5e

피드백 주신 대로 개선해보았는데, 리뷰어님이 제안해 주신 부분을 제가 잘 이행한걸까요?

igy95 commented on src/components/molecules/IconButton/index.tsx:
- `Button` 컴포넌트 안에 `Icon` 컴포넌트를 위한 스펙을 확장시키는 방향도 있을 것 같은데, `IconButton`을 별도로 만들어주신 의도가 궁금합니다!

Jxxunnn replied:
버튼은 children을 받아 유연하게 처리하려는 상황에서, 아이콘은 이미지(src) 혹은 SVG 리액트 컴포넌트로 다루게 됩니다. 그렇게 되면 SVG 관련 스타일 속성이나 이미지 속성을 제어해야할 필요가 생깁니다. 이 부분은 버튼이 공통적으로 다루는 관심사가 아니므로 아이콘 버튼을 별도로 독립시켜 이를 따로 처리하고자 했습니다.

igy95 replied:
네 그것도 틀린 말은 아니지만 앱에서 사용하는 button의 variation 범위를 따져보았을 때 `Button` 내부에서 spec을 확장하여 사용하는 방향도 고려해보시면 좋겠습니다. 현재와 같은 방향이라면 버튼의 스펙이 추가될 때마다 아래의 컴포넌트들이 추가로 생겨야 하기 때문에 중복 코드 생산 가능성이 있기 때문입니다.

- `Button`
- `IconButton`
- `LeftIconTextButton`
- `RightIconTextButton`
- ...

정답은 아니지만, 보다 구체적인 예시 차원에서 저희 회사의 [디자인 시스템](https://main--62bead1508281287d3c94d25.chromatic.com/?path=/docs/components-button--docs) 문서를 첨부하니 확인해보시면 좋겠습니다~

Jxxunnn replied:
> 앱에서 사용하는 button의 variation 범위

말씀해주신 부분을 고려하면서 UI를 재검토해보니 현재와 같은 방향에서는 직관적으로 IconButton이 필요하지 않을 것 같다는 생각이 들었습니다.

![image](https://github.com/next-step/react-payments/assets/86228307/d7f6ae44-3f08-4017-ae5f-f5d21436a70b)

추후에 이러한 UI 요구사항이 생겼을 때는 필요할 것 같다고 생각은 들지만 명확하게 설명하지는 못하겠어요. 스스로 고민해보고 설명할 수 있도록 노력해보겠습니다. 양질의 자료도 공유해주셔서 감사합니다. 살피고 고찰하면서 알아가보겠습니다! 🙏

igy95 commented on src/components/molecules/IconButton/index.tsx:
```diff
@@ -0,0 +1,17 @@
+
+export default function IconButton({ onClick, className, size, src, alt = '', iconProps }: IconButtonProps) {
+  return (
+    <button type="button" onClick={onClick} className={className} aria-label={alt}>
```
기본 `button` props를 여기에 달아줘야 하지 않을까요? prop interface에서는 확장시켜주신 거 같은데 여기에는 연결되어 있지 않아 보이네요~

Jxxunnn replied:
https://github.com/next-step/react-payments/pull/107/commits/4786731143beaa8e904af3781aa17e1f8e93ff2b

연결했습니다. 정확한 지적 감사합니다. 🙏

igy95 commented on src/components/molecules/Input/index.tsx:
```diff
@@ -0,0 +1,61 @@
+import { ComponentProps, useId } from 'react';
+import { twMerge } from 'tailwind-merge';
+
+export interface InputProps extends ComponentProps<'div'> {
```
- 이 컴포넌트는 `input`에 초점을 맞추고 있는데 input props를 확장해야 하지 않을까요? 반대로 만일 `label`과 `input`을 합성해서 사용하고자 하는 목적이었다면 컴포넌트의 이름이 `Input`보다는 좀 더 넓은 범위여야 할 것 같습니다.

Jxxunnn replied:
오 감사합니다. 너무 맞는 말인 것 같습니다. label과 input을 합성하여 사용하는  https://mui.com/material-ui/react-text-field/ 를 참고하여 이름을 TextField로 변경하였습니다.

https://github.com/next-step/react-payments/pull/107/commits/084cdd01d6d023dcdf4d40bbf52ebc5fcb03b74f

igy95 commented on src/components/molecules/Input/index.tsx:
```diff
@@ -0,0 +1,61 @@
+          </label>
+          {lengthLimit?.show && (
+            <Text size="xs">
+              {value?.toString().length}/{lengthLimit.maxLength}
```
`value`를 선언할 때 원하는 포맷으로 넣어주면 될 듯한데, `replacer` props는 왜 필요한 걸까요? 더군다나 `replacer`를 사용하면 여기서 `value?.toString().length`와 실제 입력하는 값의 길이가 다를 수도 있지 않을까요?

Jxxunnn replied:
제어 컴포넌트로 구현하면서도 props로 내리는 상태 제어 로직은 간결하게 처리하고
```
  const handleCardInputChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    setCardInputValue({ ...cardInputValue, [e.target.name]: e.target.value });
  };
```
비즈니스 로직을 각각의 도메인 컴포넌트 내부에서 처리하고 싶었는데 말씀해주신 대로 분리의 기준이 애매하고 문제의 소지가 다분한 것 같아요. 

> value를 선언할 때 원하는 포맷으로 넣어주면 될 듯한데

적용하고 hooks 패턴으로 개선하였습니다. 🙏

https://github.com/next-step/react-payments/pull/107/commits/24cc892e43e9c014b6d679b955f9a89a424f1aa8

igy95 commented on src/components/organisms/CreditCardInputs/index.tsx:
```diff
@@ -0,0 +1,135 @@
+      value={value}
+      onChange={onChange}
+      label="카드 번호"
+      lengthLimit={{ show: false, maxLength: 19 }} // 4자리마다의 '-'를 포함한 최대 길이는 19
```
`19`라는 값에 대해 상수 처리 필요하지 않을까요!

Jxxunnn replied:
https://github.com/next-step/react-payments/pull/107/commits/73ec9de0dc9d7f265dc58cc383eb84724dcd8655

넵 처리했습니다. 해당 주석을 대체할 수 있는 의미있는 상수명을 작성하는게 어렵네요. 리뷰어님이라면 어떤 이름으로 작명하셨을지 궁금합니다. 🧐

igy95 replied:
저도 크게 다르진 않을 거 같고`CARD_NUMBER_INCLUDING_HYPHEN_MAX_LENGTH`로 했을 것 같습니다. formattedCode라는 단어는, 이 모듈 외부에서 해당 상수를 사용할 때 의미를 정확하게 알 수 없을 듯 해서요.

igy95 commented on src/components/organisms/CreditCardInputs/index.tsx:
```diff
@@ -0,0 +1,135 @@
+
+  const capitalizeFirstLetter = (string: string) => {
+    return string.charAt(0).toUpperCase() + string.slice(1);
+  };
```
이건 util로 분리할 수 있어 보이네요~

Jxxunnn replied:
https://github.com/next-step/react-payments/pull/107/commits/e29bd98b0669210958ad4fc1684a6cf481b2cd09

넵 분리했습니다. 🙏

저는 사실 이렇게 도메인의 흔적이 없는 순수함수여도  재활용이 되는 상황이 오기 전까지는 공용 폴더로 빼지 않는 편인데요.
이유는 응집도를 높여 코드 추적 원활, 해당 함수를 사용하는 파일이 더이상 안쓰이게될 때 cascading delete용이해서 입니다.

리뷰어님은 바로바로 공용 폴더들이 모여있는 곳으로 옮기는 편이신가요?

igy95 replied:
> 저는 사실 이렇게 도메인의 흔적이 없는 순수함수여도 재활용이 되는 상황이 오기 전까지는 공용 폴더로 빼지 않는 편인데요.

이걸 수정하기 위해서 결국 의존해야하는 것은 '작성자의 기억'밖에 없습니다. 하지만 실무 환경에서는 협업을 하거나 개발자가 하나의 업무 컨텍스트만 가지고 일을 하지는 않기 때문에 기억에 의존한 TODO는 잊히기 마련이고 장기적으로 중복으로 선언되는 코드를 양산한다고 생각합니다.

가장 이상적으로는 분리 + 문서화겠지만, 문서화가 쉽지 않다면 적어도 분리는 제때제때하자, 가 저의 생각입니다.

igy95 commented on src/components/organisms/CreditCardInputs/index.tsx:
```diff
@@ -0,0 +1,135 @@
+  );
+}
+
+export default Object.assign(
```
`Object.assign`을 선언하신 의도가 궁금합니다!

Jxxunnn replied:
일반적으로 합성 컴포넌트를 제작할 때 메인과 서브 컴포넌트를 묶어서 export 하는 과정이 필요한데, Object.assign 빌트인 메서드를 활용하는게 가장 깔끔한 방법이라고 생각하는 편입니다. 

왜 궁금증이 생기셨는지 여쭤봐도 될까요? 메인컴포넌트를 별도로 넣지 않아서 코드가 다소 의아해보일 수도 있었을 것 같아요!

```
export const Dialog = Object.assign(DialogMain, {
  Dimmed: DialogDimmed,
  Title: DialogTitle,
  Subtitle: DialogSubtitle,
  Description: DialogDescription,
  Comment: DialogComment,
  CheckButton: DialogCheckButton,
  CheckBox: DialogCheckBox,
  TextButton: DialogTextButton,
  Button: DialogButton,
  LabelButton: DialogLabelButton,
  Divider: DialogDivider,
});
```

igy95 replied:
> 메인컴포넌트를 별도로 넣지 않아서 코드가 다소 의아해보일 수도 있었을 것 같아요!

넵 이게 이유였습니다! 현재는 그냥 아래와 같이 작성해도 무관해보여서요~

```ts
export default {
    Root,
    CardNumber,
    ExpirationDate,
    OwnerName,
    VerificationCode,
    CardPassword,
  }
```

igy95 commented on src/components/organisms/CreditCardInputs/index.tsx:
```diff
@@ -0,0 +1,135 @@
+  );
+}
+
+function Root({ children, className, ...rest }: ComponentProps<'div'>) {
```
이건 `CardInput` 영역의 책임이 아니지 않을까요? 단순 레이아웃 목적이고 `Root`를 통해서 하위 컴포넌트끼리 상태를 공유한다거나 그러진 않는 듯 해서요~

Jxxunnn replied:
정확한 지적인 것 같습니다. 불필요하고 어색하여 제거하였습니다. 🙏 

https://github.com/next-step/react-payments/pull/107/commits/be396e40ba3e529a03cbc0617a89ff26cd7fef32

igy95 commented on src/components/templates/CreateCardForm/index.tsx:
```diff
@@ -0,0 +1,67 @@
+    cardPassword: '',
+  });
+
+  const areAllInputsFilled = Object.values(cardInputValue).some((value) => value === '');
```
렌더링이 될 때마다 계산을 수행할 것 같은데, `useMemo`와 같은 렌더링 최적화 훅을 사용해보시면 어떨까요?

Jxxunnn replied:
음 해당 파일에서 로직상 렌더링을 유발하는 react state는 cardInputValue 하나 뿐이고, 해당 로직은 cardInputValue의 값을 렌더링 시마다 체크해야하기에, 현재는 렌더링시마다 계산을 수행하는 것이 기댓값입니다. 따라서 useMemo를 적용해도 기능상 변경점이 있지는 않을 것 같아요. 🧐

igy95 replied:
네 코멘트 확인해보니 그 말도 일리가 있는 것 같습니다. 저 같은 경우 컴포넌트 내에 추후 늘어날 의존성을 고려하여 데이터 계산 로직은 미리 최적화를 해두는 것을 선호하는 편이긴 한데, 이건 시점 문제라 개인의 판단에 맡겨도 괜찮을 것 같네요.

igy95 reviewed (changes requested):
안녕하세요 준근님~ 페이먼츠 미션 고생많으셨습니다 ㅎㅎ 세부 코멘트 확인 부탁드리고 질문에 대한 답변 남기고 마무리할게요!

> 아토믹 디자인 패턴에서 template과 page를 어떻게 구분짓는 것이 적절할까요?

현업 경험이 길다고 할 수는 없지만 그래도 어느정도 해오면서 실무에서 아토믹 디자인 패턴을 사용한 적은 없었던 것 같습니다. 왜 그럴까 생각해보면 이 패턴에서 표방하는 개념과 실제 업무에 사용하려는 목적 사이에 괴리가 커서 그런게 아닌가.. 싶네요.

사람마다 각 단계에 들어가야하는 컴포넌트에 대한 생각이 다르기도 하고 컴포넌트를 4~5단계로 나눠야할 만큼 관리 포인트가 복잡한가 고려해보면 실효성이 떨어지는 것 같아요. 그래서 획일화된 패턴을 사용하기보다는, 단순 ui를 표현하는 `components`, 다른 레이어와 교류하는 `containers`을 기본 전제로 두고 필요에 따라 레이어를 추가하는 식으로 유연하게 구조를 가져가는 편입니다.

> CDD란 무엇인가요?

CDD가 무엇인가에 대해서는 제 생각보다는 관련 아티클이나 소개글을 찾아보시는 게 좋을 듯 합니다. 그리고 이에 대한 장점이라 함은 결국 관심사의 분리 + 작은 문제부터 집중하여 큰 문제를 해결할 수 있는 흐름이 아닐까요? 딱 들어맞진 않을 수도 있지만, 실무 관점으로는 각 회사의 디자인 시스템을 하나의 예시로 볼 수 있겠네요.

> 리뷰는 어떻게 하는 것이 좋을까요?

저는 이 두가지가 가장 중요하지 않나 싶네요.

1. 하나의 PR은 (되도록 작은 단위에서) 하나의 문제만 해결하도록 한다.
2. 코드 상으로만 알 수 있는 것과 코드로는 알 수 없는 것들을 구분할 수 있어야 한다. 코드로 알 수 없는 것들에 대해 리뷰를 요청한 상황이라면 코멘트나 주석을 통해 먼저 알려줄 수 있어야 한다.

다른 무엇보다 내 코드를 읽어볼 상대방 입장에서 이 사람이 어떤 생각을 할까 고민해보면 좋을 것 같네요.

> 스토리북을 평소 어떻게 활용하시나요?

잘 만들어진 테스트는 그 자체로 명세서가 될 수 있다고 생각하고 스토리북도 그 중 하나라고 봅니다. 명세의 장점으로 아래와 같은 상황에서 이점을 얻을 수 있겠죠.

1. 디자이너와 소통할 때
2. 다른 사람이 만든 컴포넌트의 사용례를 찾고자 할 때

> Funnel 기반으로 작성될 어플리케이션에서는 무엇을 고려해야할까요?

질문 범위가 너무 넓은 것 같습니다. 단순히 정답을 위한 질문을 던지기 전에 고민의 범위를 좁히기 위해 먼저 관련된 글들을 찾아보시는 걸 권장 드리고 싶네요.

> 제어 컴포넌트, 비제어 컴포넌트

리액트가 data를 핸들링하여 화면을 그리기 위해 지향하는 flow로 구현 상 한계가 존재하는 상황에서는 비제어 컴포넌트를 사용합니다. 제 생각엔 주로 화면 상의 요소를 다룰 때인 것 같아요.

- 스크롤 위치를 파악해서 어떤 액션을 취할 때
- form 의 포커스를 옮길 때
- 사용자 브라우저의 width나 height를 가져와서 액션을 취할 때

Jxxunnn committed:
chore: update readme

Jxxunnn committed:
chore: rename storybook readme

Jxxunnn committed:
refactor: remove unused routes

Jxxunnn committed:
refactor: rename Box component

Jxxunnn committed:
chore: remove comment

Jxxunnn committed:
chore: remove useless ignore lines

Jxxunnn committed:
refactor: separate and combine typography type

Jxxunnn committed:
feat: pass html button type as rest props

Jxxunnn committed:
chore: config absolute path

Jxxunnn committed:
refactor: move the function to the common utility

Jxxunnn committed:
refactor: replace magic number with constant

Jxxunnn committed:
refactor: remove 'Root' subcomponent

Jxxunnn committed:
refactor: rename 'Input' to 'TextField'

label과 input을 합성하여 사용하는 컴포넌트이기에 primitive ui로 착각할 여지가 있어 이름을 변경합니다.

Jxxunnn committed:
refactor: convert replacer prop to hooks pattern

- UI data와 state의 동기화를 위해 TextField 컴포넌트에서 replacer prop 제거
- replacer prop을 사용하는 컴포넌트를 Hooks Pattern으로 개선
- withReplacer 스토리 제거

igy95 reviewed (approved):
안녕하세요 준근님! 페이먼츠 미션 고생 많으셨습니다 :) 추가 세부 코멘트 확인 부탁드리고 다음 미션에서 봬요~

Jxxunnn commented:
광열님, 리뷰 진심으로 감사드립니다. 여러 피드백을 통해 많은 것을 배우고 성장하는 소중한 시간이었습니다. 이렇게 코드와 글로 소통하게 되어 너무 영광이었고, 다음 미션에서 더 좋은 모습으로 찾아뵙도록 노력하겠습니다. 🙏
